### PR TITLE
Check forbidden kernel constructs in constant evaluated context

### PR DESCRIPTION
**Pull request by keryell**
There are a lot of restrictions for kernel code, except when present in a constant evaluated context.

This is a test corresponding to the specification clarification introduced by PR https://github.com/KhronosGroup/SYCL-Docs/pull/388

### CODE REVIEW

File: tests/language/constant_evaluation.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      // A constant-evaluated recursive function.
+      constexpr auto result = f(10);
+      a[0] = result;
+
+      if constexpr (false) {
+        // A non constant-evaluated recursive function going further more
+        // through forbidden function pointer which is skipped by the if
+        // constexpr.
+        auto* p = f;
+        auto other = p(5) + f(6);
```
Comment (gmlueck) [inline tests/language/constant_evaluation.cpp:52]: I think that this part does not test a feature used in a [manifestly constant evaluated expression](https://en.cppreference.com/w/cpp/language/constant_expression#Manifestly_constant-evaluated_expressions).

Rather, I think this is testing that you can use forbidden C++ features in a [discarded statement](https://en.cppreference.com/w/cpp/language/if#Constexpr_if).  We did not make this legal in KhronosGroup/SYCL-Docs#388.  Potentially, we could consider making this legal, but that would require a separate change to the spec.

Comment (keryell) [reply tests/language/constant_evaluation.cpp:52]: I have allowed "discarded statements" in [KhronosGroup/SYCL-Docs@`966a051` (#388)](https://github.com/KhronosGroup/SYCL-Docs/pull/388/commits/966a051946f1a3f8e4880cee9b5a4f828420bfea)

File: tests/language/constant_evaluation.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// A very inefficient way to compute Fibonacci's series using recursion which is
+// forbidden inside non constant-evaluated kernel code/
+auto constexpr f(int v) {
+  if (v < 2) return v;
+  return f(v - 1) + f(v - 2);
+}
+
+// This corresponds to clarification introduced with
+// https://github.com/KhronosGroup/SYCL-Docs/pull/388
+TEST_CASE("constant evaluated code works inside kernel", "[language]") {
```
Comment (bader) [inline tests/language/constant_evaluation.cpp:38]: @keryell, please, disable this test case for DPC++ implementation.

I see following errors in CI:
```
/__w/SYCL-CTS/SYCL-CTS/tests/language/constant_evaluation.cpp:33:10: error: SYCL kernel cannot call a recursive function
   33 |   return f(v - 1) + f(v - 2);
      |          ^
/__w/SYCL-CTS/SYCL-CTS/tests/language/constant_evaluation.cpp:31:1: note: function implemented using recursion declared here
   31 | auto constexpr f(int v) {
      | ^
/__w/SYCL-CTS/SYCL-CTS/tests/language/constant_evaluation.cpp:33:21: error: SYCL kernel cannot call a recursive function
   33 |   return f(v - 1) + f(v - 2);
      |                     ^
/__w/SYCL-CTS/SYCL-CTS/tests/language/constant_evaluation.cpp:31:1: note: function implemented using recursion declared here
   31 | auto constexpr f(int v) {
      | ^
2 errors generated.
```

**Comment by tomdeakin** [comment]
https://github.com/KhronosGroup/SYCL-Docs/pull/388 was approved to merge. 
This is just waiting the temporary disable.

File: tests/language/constant_evaluation.cpp (hunk 2 above)
Comment (steffenlarsen) [reply tests/language/constant_evaluation.cpp:38]: https://github.com/KhronosGroup/SYCL-CTS/pull/987 version bump should hopefully address this issue.

**Comment by steffenlarsen** [comment]
With https://github.com/KhronosGroup/SYCL-CTS/pull/987 merged, the DPC++ CI should be passing. @keryell - I took the liberty of updating the branch. Be aware of the need for a local pull. 😄

**Review by gmlueck — APPROVED** [review]

**Pull request merged**